stan/math/fwd/fun/multiply.hpp

```diff
@@ -10,120 +11,64 @@
 namespace stan {
 namespace math {
 
-template <typename T, int R1, int C1>
-inline Eigen::Matrix<fvar<T>, R1, C1> multiply(
-    const Eigen::Matrix<fvar<T>, R1, C1>& m, const fvar<T>& c) {
-  Eigen::Matrix<fvar<T>, R1, C1> res(m.rows(), m.cols());
-  for (int i = 0; i < m.size(); i++)
-    res(i) = c * m(i);
-  return res;
-}
-
-template <typename T, int R2, int C2>
-inline Eigen::Matrix<fvar<T>, R2, C2> multiply(
-    const Eigen::Matrix<fvar<T>, R2, C2>& m, double c) {
-  Eigen::Matrix<fvar<T>, R2, C2> res(m.rows(), m.cols());
-  for (int i = 0; i < m.size(); i++)
-    res(i) = c * m(i);
-  return res;
-}
-
-template <typename T, int R1, int C1>
-inline Eigen::Matrix<fvar<T>, R1, C1> multiply(
-    const Eigen::Matrix<double, R1, C1>& m, const fvar<T>& c) {
-  Eigen::Matrix<fvar<T>, R1, C1> res(m.rows(), m.cols());
-  for (int i = 0; i < m.size(); i++)
-    res(i) = c * m(i);
-  return res;
-}
-
-template <typename T, int R1, int C1>
-inline Eigen::Matrix<fvar<T>, R1, C1> multiply(
-    const fvar<T>& c, const Eigen::Matrix<fvar<T>, R1, C1>& m) {
-  return multiply(m, c);
-}
-
-template <typename T, int R1, int C1>
-inline Eigen::Matrix<fvar<T>, R1, C1> multiply(
-    double c, const Eigen::Matrix<fvar<T>, R1, C1>& m) {
-  return multiply(m, c);
-}
-
-template <typename T, int R1, int C1>
-inline Eigen::Matrix<fvar<T>, R1, C1> multiply(
-    const fvar<T>& c, const Eigen::Matrix<double, R1, C1>& m) {
-  return multiply(m, c);
-}
-
-template <typename T, int R1, int C1, int R2, int C2>
-inline Eigen::Matrix<fvar<T>, R1, C2> multiply(
-    const Eigen::Matrix<fvar<T>, R1, C1>& m1,
-    const Eigen::Matrix<fvar<T>, R2, C2>& m2) {
+template <
+    typename T1, typename T2, typename = require_all_eigen_vt<is_fvar, T1, T2>,
+    typename = require_same_t<typename value_type_t<T1>::Scalar,
+                              typename value_type_t<T2>::Scalar>,
+    typename
+    = require_t<bool_constant<T1::ColsAtCompileTime == T2::RowsAtCompileTime>>,
+    typename = require_not_t<
+        conjunction<is_eigen_row_vector<T1>, is_eigen_col_vector<T2>>>,
+    int = 0>
+inline auto multiply(const T1& m1, const T2& m2) {
   check_multiplicable("multiply", "m1", m1, "m2", m2);
-  Eigen::Matrix<fvar<T>, R1, C2> result(m1.rows(), m2.cols());
-  for (size_type i = 0; i < m1.rows(); i++) {
-    Eigen::Matrix<fvar<T>, 1, C1> crow = m1.row(i);
-    for (size_type j = 0; j < m2.cols(); j++) {
-      Eigen::Matrix<fvar<T>, R2, 1> ccol = m2.col(j);
-      result(i, j) = dot_product(crow, ccol);
-    }
-  }
-  return result;
+  return m1 * m2;
```

Comment:
Is a `.cast<T>` when the matrix already has scalar type T a no-op? If so maybe one thing could do here is `m1.cast<return_type_t<T1, T2>>() * m2.cast<return_type_t<T1, T2>>();` or something

Reply:
Yep, that's a no-op. Got an example on godbolt here where it compiles down to the same assembly [with a cast](https://godbolt.org/z/flQN9u) and [without a cast](https://godbolt.org/z/CnbEev), for a simple assign, at least

Reply:
Unfortunately that being a no-op is not enough to guarantee speedup here. The problem is that matrix multiplication can not be combined with (most) other operations, so if cast is not a no-op it results in a copy of the matrix. 

I benchmarked this. For two square matrices it is comparable to existing code. If the matrix that is being typecast is significantly larger it gets slower.

Reply:
I am not sure if this can be reproduced with just built-in types. What I had in mind is:
```
Eigen::MatrixXd a;
Eigen::Matrix<fvar, Eigen::Dynamic, Eigen::Dynamic> b;
Eigen::Matrix<fvar, Eigen::Dynamic, Eigen::Dynamic> c = a.cast<fvar>() * b;
```